sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v2_7/q14a/explain.txt

```diff
@@ -1,129 +1,168 @@
+            :        :                 :     :                          +- ReusedExchange (42)
+            :        :                 :     +- BroadcastExchange (57)
+            :        :                 :        +- * BroadcastHashJoin LeftSemi BuildRight (56)
+            :        :                 :           :- * Filter (54)
+            :        :                 :           :  +- * ColumnarToRow (53)
+            :        :                 :           :     +- Scan parquet spark_catalog.default.item (52)
+            :        :                 :           +- ReusedExchange (55)
+            :        :                 +- ReusedExchange (60)
+            :        :- * Filter (81)
+            :        :  +- * HashAggregate (80)
+            :        :     +- Exchange (79)
+            :        :        +- * HashAggregate (78)
+            :        :           +- * Project (77)
+            :        :              +- * BroadcastHashJoin Inner BuildRight (76)
+            :        :                 :- * Project (74)
+            :        :                 :  +- * BroadcastHashJoin Inner BuildRight (73)
+            :        :                 :     :- * BroadcastHashJoin LeftSemi BuildRight (71)
+            :        :                 :     :  :- * Filter (69)
+            :        :                 :     :  :  +- * ColumnarToRow (68)
+            :        :                 :     :  :     +- Scan parquet spark_catalog.default.catalog_sales (67)
+            :        :                 :     :  +- ReusedExchange (70)
+            :        :                 :     +- ReusedExchange (72)
+            :        :                 +- ReusedExchange (75)
+            :        +- * Filter (96)
+            :           +- * HashAggregate (95)
+            :              +- Exchange (94)
+            :                 +- * HashAggregate (93)
+            :                    +- * Project (92)
+            :                       +- * BroadcastHashJoin Inner BuildRight (91)
+            :                          :- * Project (89)
+            :                          :  +- * BroadcastHashJoin Inner BuildRight (88)
+            :                          :     :- * BroadcastHashJoin LeftSemi BuildRight (86)
+            :                          :     :  :- * Filter (84)
+            :                          :     :  :  +- * ColumnarToRow (83)
+            :                          :     :  :     +- Scan parquet spark_catalog.default.web_sales (82)
+            :                          :     :  +- ReusedExchange (85)
+            :                          :     +- ReusedExchange (87)
+            :                          +- ReusedExchange (90)
+            :- * HashAggregate (114)
+            :  +- Exchange (113)
+            :     +- * HashAggregate (112)
+            :        +- * HashAggregate (111)
+            :           +- * HashAggregate (110)
+            :              +- Union (109)
```

Comment:
Ouch, this might cause performance regression as we no longer have `Exchange (99)` which we could reuse.

Reply:
Hm, interesting. I think it may depend on the cost of the removed `Exchange (99)` and if it can cover the segments:

```
* HashAggregate
+- Union
      :- * Filter
      :  +- * HashAggregate
      :- * Filter
      :  +- * HashAggregate
      :- * Filter
      :  +- * HashAggregate
```

Reply:
Yeah. Unfortunately the above happens 4 times in this query.
But still, IMO this PR is a step in the right direction to remove unnecessary exchanges.
In the future some kind of common subplan elimination rule should identify and cache query parts that appear multiple times if computation costs exceed the cost of caching.